netket/experimental/operator/_fermion_operator_2nd_jax.py

```diff
@@ -0,0 +1,686 @@
+    xp_u = jnp.take_along_axis(xp_padded, i_nonzero[..., None], axis=-2)
+    mels_u = jnp.take_along_axis(mels_padded, i_nonzero, axis=-1)
+
+    # TODO here would be the place to remove / merge repeated mels
+    #
+    # you should check that n_nonzero <= max_conn_size outside of jit,
+    # and increase max_conn_size if it's not
+    return xp_u, mels_u, n_nonzero
+
+
+@partial(jax.jit, static_argnums=(0, 4))
+def n_conn_jax(dtype, tl_diag, tl_offdiag, x, apply_terms_fun=apply_terms_scan):
+    max_conn_size = 0
+    # let dce take care of not computing xp
+    _, _, n_conn = get_conn_padded_jax(
+        max_conn_size,
+        dtype,
+        tl_diag,
+        tl_offdiag,
+        x,
+        apply_terms_fun=partial(apply_terms_scan, unroll=4),
+    )
+    return n_conn
+
+
+@register_pytree_node_class
+class FermionOperator2ndJax(FermionOperator2ndBase, DiscreteJaxOperator):
+    r"""
+    A fermionic operator in :math:`2^{nd}` quantization using pure jax dark
+    magic for indexing.
+
+    .. warning::
+
+        This class is not a Pytree, so it **can** be used inside of jax-transformed
+        functions like `jax.grad` or `jax.jit`.
+
+    When using native (experimental) sharding, or when working with GPUs,
+    we reccomend using this implementations of the operators for potentially
+    better performance.
+
+    Can be converted to the Numba one with the method
+    :meth:`~netket.experimental.operator.FermionOperator2ndJax.to_numba_operator()`.
+    """
+
+    def _setup(self, force: bool = False):
+        """Analyze the operator strings and precompute arrays for get_conn inference"""
+        if force or not self._initialized:
+            # TODO ideally we would set dagger_dtype to the same as x
+            # however, unfortunately, the dtype of the states in netket
+            # is stored in the sampler and not in hilbert, so we don't know it at this stage
+            diag_operators = {
+                k: v for k, v in self._operators.items() if _is_diag_term(k)
+            }
+            offdiag_operators = {
+                k: v for k, v in self._operators.items() if not _is_diag_term(k)
+            }
+
+            self._terms_list_diag = prepare_terms_list(
+                diag_operators,
+                self._constant,
+                site_dtype=np.uint32,
+                dagger_dtype=jnp.bool_,
+                weight_dtype=self._dtype,
+            )
+            self._terms_list_offdiag = prepare_terms_list(
+                offdiag_operators,
+                site_dtype=np.uint32,
+                dagger_dtype=jnp.bool_,
+                weight_dtype=self._dtype,
+            )
+
+            # TODO the following could be reduced further
+            self._max_conn_size = int(len(self._terms_list_diag) > 0) + len(
+                offdiag_operators
+            )
+            self._initialized = True
+
+            self._kwargs = {"apply_terms_fun": apply_terms_scan_bits}
```

Comment:
I had introduced this field is as a hacky way to set the implementation. 
Maybe we could add the constructor to this class with a private parameter for it, similar to `_mode` in paulistringsjax.

IIrc apply_terms_scan_bits was the fastest implementation, which is why it is set as default here.
(this is memory bound, so doing the computation with bitmasks helps)
One day we should do some proper benchmarking...

Also I think the one with masks might be useful for people which want to change their parameters without recompiling.

Reply:
the implementation is `apply_terms_masks`or `apply_terms_only_masks`?